src/projection.js

```diff
@@ -0,0 +1,95 @@
+const {acquireAsyncIterator} = require('./iterator.js')
+const {acquireSessionLock, inTransaction, releaseSessionLock} = require('./pg.js')
+const {allSerial} = require('./async.js')
+const {readEventsContinuously} = require('./event.js')
+const {systemClock} = require('./clock.js')
+
+module.exports = {
+  maintainProjection,
+}
+
+function maintainProjection (pgPool, name, apply, options = {}) {
+  const {timeout = 100, clock = systemClock} = options
+  const iterator = createProjectionIterator(pgPool, name, apply, timeout, clock)
+
+  return {
+    [Symbol.asyncIterator]: () => iterator,
+    cancel: iterator.cancel,
+  }
+}
+
+function createProjectionIterator (pgPool, name, apply, timeout, clock) {
+  const lockName = `projection.${name}`
+  let offset, iterator, pgClient
+  let isLocked = false
+
+  return {
+    async next () {
+      if (!iterator) {
+        pgClient = await pgPool.connect()
+
+        await acquireSessionLock(pgClient, lockName)
```

Comment:
As discussed, it's probably worth the effort to use a sequence ID on the projections offset table as the lock ID, to avoid any chance of a hash collision, however unlikely.

Reply:
It is done. Turns out it adds more queries than I thought, because you need to query for the ID, then acquire the lock, then query for the next offset. No big deal though.